plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -132,49 +132,72 @@ public function add_navigation_id_directive( $block_content, $block, $instance )
+
+		// Only proceed if the block is a product collection block and query IDs match.
+		if ( $is_product_collection_block && $query_id === $parsed_query_id ) {
+			$block_content = $this->process_pagination_links( $block_content );
+		}
 
-		if (
-			$is_product_collection_block &&
-			$instance->context['queryId'] === $this->parsed_block['attrs']['queryId']
-		) {
-			$p = new \WP_HTML_Tag_Processor( $block_content );
-			$p->next_tag( array( 'class_name' => 'wp-block-query-pagination' ) );
+		return $block_content;
+	}
 
-			while ( $p->next_tag( 'a' ) ) {
-				$class_attr = $p->get_attribute( 'class' );
-				$class_list = preg_split( '/\s+/', $class_attr );
+	/**
+	 * Process pagination links within the block content.
+	 *
+	 * @param string $block_content The block content.
+	 * @return string The updated block content.
+	 */
+	private function process_pagination_links( $block_content ) {
+		$p = new \WP_HTML_Tag_Processor( $block_content );
+		$p->next_tag( array( 'class_name' => 'wp-block-query-pagination' ) );
 
-				$is_previous         = in_array( 'wp-block-query-pagination-previous', $class_list, true );
-				$is_next             = in_array( 'wp-block-query-pagination-next', $class_list, true );
-				$is_previous_or_next = $is_previous || $is_next;
+		// This will help us to find the start of the block content using the `seek` method.
+		$p->set_bookmark( 'start' );
 
-				$navigation_link_payload = array(
-					'prefetch' => $is_previous_or_next,
-					'scroll'   => false,
-				);
+		$this->update_pagination_anchors( $p, 'page-numbers', 'product-collection-pagination-numbers' );
+		$this->update_pagination_anchors( $p, 'wp-block-query-pagination-next', 'product-collection-pagination--next' );
+		$this->update_pagination_anchors( $p, 'wp-block-query-pagination-previous', 'product-collection-pagination--previous' );
 
-				$p->set_attribute(
-					'data-wc-navigation-link',
-					wp_json_encode( $navigation_link_payload )
-				);
+		return $p->get_updated_html();
+	}
 
-				if ( $is_previous ) {
-					$p->set_attribute( 'key', 'pagination-previous' );
-				} elseif ( $is_next ) {
-					$p->set_attribute( 'key', 'pagination-next' );
-				}
+	/**
+	 * Sets up data attributes required for interactivity and client-side navigation.
+	 *
+	 * @param \WP_HTML_Tag_Processor $processor The HTML tag processor.
+	 * @param string                 $class_name The class name of the anchor tags.
+	 * @param string                 $key_prefix The prefix for the data-wc-key attribute.
+	 */
+	private function update_pagination_anchors( $processor, $class_name, $key_prefix ) {
+		// Start from the beginning of the block content.
+		$processor->seek( 'start' );
+
+		while ( $processor->next_tag(
+			array(
+				'tag_name'   => 'a',
+				'class_name' => $class_name,
+			)
+		) ) {
+			$processor->set_attribute( 'data-wc-interactive', '{ "namespace": "woocommerce/product-collection" }' );
```

Comment:
Isn't this a duplication of what we do in `add_navigation_id_directive` method https://github.com/woocommerce/woocommerce/pull/43817/files#diff-933825fdae3edb1301562e70acef6627856dd1c39c7475056ef21ac4f0ee9941R116? Or the other way around, duplicating there what we do in here?

Reply:
The reason was that if we don’t explicitly add a namespace using the following line:

```jsx
$processor->set_attribute( 'data-wc-interactive', '{ "namespace": "woocommerce/product-collection" }' );
```

We need to explicitly add namespace with each callback & action, for example:

```jsx
$processor->set_attribute( 'data-wc-watch', 'woocommerce/product-collection::callbacks.prefetch' );
```

@DAreRodz @luisherranz Is there any reason why we need to explicitly add namespace with actions and callbacks if parent (Product Collection) block have defined namespace? AFAIU It should automatically inherit namespace from it’s parent 🤔

Reply:
When I decided to include the namespace explicitly on the directive values, my reasoning was that, as they were placed in a different block, another interactive block could be placed in between and overwrite the namespace, thus making the `data-wc-watch` directives look for the `callbacks.prefetch` in the wrong namespace.

Another way to prevent this would be to add a `data-wc-interactive` directive with the required namespace, but I wanted to try the other syntax. 🙂

Reply:
> another interactive block could be placed in between and overwrite the namespace


For me, there was no interactive block in-between. There was just Product Collection block. I am not sure why following code wasn’t working:

```jsx
$processor->set_attribute( 'data-wc-watch', 'callbacks.prefetch' );
```

If there is a parent block (Product Collection) with namespace `woocommerce/product-collection`, shouldn’t it use “callbacks.prefetch” from this namespace? I tried it, and it wasn’t working.

Reply:
> shouldn’t it use “callbacks.prefetch” from this namespace?

It should! That's interesting. Could you check that there isn't any `data-wc-interactive` in between with a different namespace in the final HTML? That's the only thing I can think of that could break it. 🤔

Reply:
@DAreRodz, I've just recorded a video to explain the issue, as I thought it would be more effective than a written explanation. Please correct me if I misunderstood how namespace should be working 🙂

https://github.com/woocommerce/woocommerce/assets/16707866/074090ec-10ee-443f-9936-e20ff4f129b2